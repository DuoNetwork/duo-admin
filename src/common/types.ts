import { ChildProcess } from 'child_process';
export * from '../../../duo-contract-wrapper/src/types';

export interface IBaseMarketData {
	source: string;
	base: string;
	quote: string;
	timestamp: number;
}

export interface ITrade extends IBaseMarketData {
	id: string;
	price: number;
	amount: number;
}

export interface IPriceFix extends IBaseMarketData {
	price: number;
	volume: number;
}

export interface IPrice extends IBaseMarketData {
	period: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
}

export interface IBaseEvent {
	contractAddress: string;
	timestamp: number;
	transactionHash: string;
	blockNumber: number;
}

export interface IAcceptedPrice extends IBaseEvent {
	price: number;
	navA: number;
	navB: number;
}

export interface ITotalSupply extends IBaseEvent {
	tokenA: number;
	tokenB: number;
}

export interface IConversion extends IBaseEvent {
	type: string;
	eth: number;
	tokenA: number;
	tokenB: number;
	fee: number;
	pending?: boolean;
	reverted?: boolean;
}

export interface IStatus {
	process: string;
	timestamp: number;
}

export interface IPriceStatus extends IStatus {
	price: number;
	volume: number;
}

export interface INodeStatus extends IStatus {
	block: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	forceREST: boolean;
	pair: string;
	sources: string[];
	exSources: string[];
	assets: string[];
	live: boolean;
	dbLive: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	source: string;
	event: string;
	provider: string;
	period: number;
	base: string;
	quote: string;
	contractType: string;
	tenor: string;
}

export interface ISourceSettings {
	priceInversed: boolean;
	quoteInversed: boolean;
	separator: string;
	isLowercase: boolean;
	tradesInterval: number;
	supportWS: boolean;
	filterByTimestamp: boolean;
	wsLink: string;
}

export interface ISourceAsset {
	base: {
		[code: string]: boolean;
	};
	quote: {
		[code: string]: boolean;
	};
	mapping: string;
	name: string;
}
export interface ISource {
	assets: {
		[code: string]: ISourceAsset;
	};
	settings: ISourceSettings;
}

export interface IBitfinexRawTradeWS {
	timestamp: number;
	seq: string;
	id: number;
	price: string;
	amount: string;
}

export interface IBitstampRawTradeRest {
	date: number;
	tid: number;
	price: number;
	type: string;
	amount: number;
}

export interface IBitstampRawTradeWs {
	microtimestamp: string;
	amount: number;
	buy_order_id: number;
	sell_order_id: number;
	amount_str: string;
	price_str: string;
	timestamp: string;
	price: number;
	type: number;
	id: number;
}

export interface IBitfinexRawTradeRest {
	timestamp: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeRest {
	timestamp: number;
	timestampms: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeWs {
	type: string;
	eventid: number;
	socket_sequence: number;
	timestamp: number;
	timestampms: number;
	events: IGeminiTradeData[];
}

export interface IGeminiTradeData {
	type: string;
	tid: number;
	price: string;
	amount: string;
	makerSide: string;
}

export type IKrakenRawTrade = IKrakenRawTradeEntry & {
	last: string;
};

export interface IKrakenRawTradeEntry {
	[pair: string]: string[][];
}

export interface IGdaxRawTradeRest {
	error: string[];
	result: {
		[code: string]: Array<Array<string | number>>;
	};
}

export interface ISubProcess {
	source: string;
	instance: ChildProcess;
	lastFailTimestamp: number;
	failCount: number;
}
