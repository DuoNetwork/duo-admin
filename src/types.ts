export interface ITrade {
	source: string;
	id: string;
	price: number;
	amount: number;
	timestamp: number;
}

export interface IPrice {
	price: number;
	volume: number;
	timestamp: number;
}

export interface IPriceBar {
	source: string;
	date: string;
	hour: string;
	minute: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
	timestamp: number;
}

export interface IEvent {
	type: string;
	id: string;
	blockHash: string;
	blockNumber: number;
	transactionHash: string;
	logStatus: string;
	parameters: {[key: string]: string};
	timestamp: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	live: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	eth: number;
	address: string;
	addr1: string;
	addr2: string;
	privateKey: string;
	price: number;
	source: string;
	pwd: string;
	event: string;
	provider: string;
	contractState: string;
	accountNum: number;
	saveAccount: boolean;
	from: string;
	to: string;
	value: number;
	index: number;
	total: number;
	minEther: number;
	alpha: number;
	amtA: number;
	amtB: number;
	numOfMinutes: number;
	numOfHours: number;
	key: string;
	endBlk: number;
}
