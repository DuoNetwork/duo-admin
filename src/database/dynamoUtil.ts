import AWS from 'aws-sdk';
import { PutItemInput } from 'aws-sdk/clients/dynamodb';
import moment from 'moment';
import * as CST from '../constants';
import { IPrice, IPriceBar, ITrade } from '../types';
import util from '../util';

export class DynamoUtil {
	private ddb: undefined | AWS.DynamoDB = undefined;
	// private role: string = '';
	private process: string = 'UNKNOWN';
	private live: boolean = false;
	public init(live: boolean, role: string, process: string) {
		this.live = live;
		this.process = process;
		// this.role = role;
		AWS.config.loadFromPath('./src/keys/aws/' + (live ? 'live' : 'dev') + '/' + role + '.json');
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public insertData(params: PutItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.putItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public convertTradeToSchema(trade: ITrade, systime: number) {
		return {
			[CST.DB_TX_ID]: { S: trade.id },
			[CST.DB_TX_PRICE]: { N: trade.price.toString() },
			[CST.DB_TX_AMOUNT]: { N: trade.amount.toString() },
			[CST.DB_TX_TS]: { N: trade.timestamp + '' },
			[CST.DB_TX_SYSTIME]: { N: systime + '' }
		};
	}

	public convertPriceToSchema(price: IPrice) {
		return {
			[CST.DB_HISTORY_PRICE]: { N: price.price + '' },
			[CST.DB_HISTORY_TIMESTAMP]: { N: price.timestamp + '' },
			[CST.DB_HISTORY_VOLUME]: { N: price.volume + '' }
		};
	}

	public convertPriceBarToSchema(priceBar: IPriceBar) {
		return {
			[CST.DB_HR_SRC_DATE]: { S: priceBar.source + '|' + priceBar.date },
			[CST.DB_HR_HOUR]: { N: priceBar.hour + '' },
			[CST.DB_HR_OPEN]: { N: priceBar.open + '' },
			[CST.DB_HR_HIGH]: { N: priceBar.high + '' },
			[CST.DB_HR_LOW]: { N: priceBar.low + '' },
			[CST.DB_HR_CLOSE]: { N: priceBar.close + '' },
			[CST.DB_HR_VOLUME]: { N: priceBar.volume + '' },
			[CST.DB_HR_TS]: { N: priceBar.timestamp + '' }
		};
	}

	public async insertTradeData(trade: ITrade, insertStatus: boolean): Promise<void> {
		const systemTimestamp = util.getNowTimestamp(); // record down the MTS
		const data = this.convertTradeToSchema(trade, systemTimestamp);

		const params = {
			TableName: this.live ? CST.DB_AWS_TRADES_LIVE : CST.DB_AWS_TRADES_DEV,
			Item: {
				[CST.DB_TX_SRC_DATE]: {
					S: trade.source + '|' + moment.utc(trade.timestamp).format('YYYY-MM-DD')
				},
				...data
			}
		};

		await this.insertData(params);
		if (insertStatus) await this.insertStatusData(data);
	}

	public async insertHourlyData(priceBar: IPriceBar): Promise<void> {
		return this.insertData({
			TableName: this.live ? CST.DB_AWS_HOURLY_LIVE : CST.DB_AWS_HOURLY_DEV,
			Item: {
				...this.convertPriceBarToSchema(priceBar)
			}
		});
	}

	public insertHeartbeat(): Promise<void> {
		return this.insertData({
			TableName: this.live ? CST.DB_AWS_STATUS_LIVE : CST.DB_AWS_STATUS_DEV,
			Item: {
				[CST.DB_ST_PROCESS]: {
					S: this.process
				},
				[CST.DB_ST_TS]: { N: util.getNowTimestamp() + '' }
			}
		});
	}

	public insertStatusData(data: object): Promise<void> {
		return this.insertData({
			TableName: this.live ? CST.DB_AWS_STATUS_LIVE : CST.DB_AWS_STATUS_DEV,
			Item: {
				[CST.DB_ST_PROCESS]: {
					S: this.process
				},
				...data
			}
		});
	}

	public readTradeData(source: string, date: string): Promise<any> {
		const params = {
			TableName: this.live ? CST.DB_AWS_TRADES_LIVE : CST.DB_AWS_TRADES_DEV,
			KeyConditionExpression: CST.DB_TX_SRC_DATE + ' = :' + CST.DB_TX_SRC_DATE,
			ExpressionAttributeValues: {
				[':' + CST.DB_TX_SRC_DATE]: { S: source + '|' + date }
			}
		};

		console.log(params);

		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.query(params, (err, data) => (err ? reject(err) : resolve(data)))
					: reject('dynamo db connection is not initialized')
		);
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
