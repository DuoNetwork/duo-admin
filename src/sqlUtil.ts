import * as mysql from 'mysql';
import * as CST from './constants';
// const math = require("mathjs");
import { Price, Trade } from './types';

export class SqlUtil {
	conn: undefined | mysql.Connection = undefined;

	init(user: string, pwd: string) {
		this.conn = mysql.createConnection({
			host: CST.DB_HOST,
			user: user,
			password: pwd,
			database: CST.DB_PRICEFEED
		});

		this.conn.connect(function(err) {
			if (err) throw err;
			console.log('Connected!');
		});
	}

	executeQuery(sqlQuery: string): Promise<any> {
		console.log(sqlQuery);
		return new Promise((resolve, reject) => {
			if (this.conn)
				this.conn.query(sqlQuery, (err, result) => {
					if (err && err.code != undefined && err.code === 'ER_DUP_ENTRY')
						// console.log('.');
						// rocess.stdout.write(".");
						reject(err);
					else if (err) reject(err);
					else resolve(result);
				});
			else reject('db connection is not initialized');
		});
	}

	async insertSourceData(sourceData: Trade) {
		const systemTimestamp = Math.floor(Date.now()); // record down the MTS

		// let price_str = math.format(price, { exponential: { lower: 1e-100, upper: 1e100 } });
		// let amount_str = math.format(amount, { exponential: { lower: 1e-100, upper: 1e100 } });

		const priceStr = sourceData.price.toString();
		const amountStr = sourceData.amount.toString();

		const sql =
			'REPLACE ' +
			CST.DB_TABLE_TRADE +
			" VALUES ('" +
			sourceData.source +
			"','" +
			sourceData.id +
			"','" +
			priceStr +
			"','" +
			amountStr +
			"','" +
			(sourceData.timestamp || systemTimestamp) +
			"','" +
			systemTimestamp +
			"')";
		console.log(await this.executeQuery(sql));
		// await this.executeQuery(sql);
	}

	async insertPrice(price: Price) {
		console.log(
			await this.executeQuery(
				'INSERT INTO ' +
					CST.DB_TABLE_HISTORY +
					" VALUES ('" +
					price.timestamp +
					"','" +
					price.price +
					"','" +
					price.volume +
					"')"
			)
		);
	}

	async readLastPrice(): Promise<Price> {
		const res = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_TABLE_HISTORY +
				' order by ' +
				CST.DB_HISTORY_TIMESTAMP +
				' DESC LIMIT 1'
		);
		return res[0]
			? {
					price: Number(res[0][CST.DB_HISTORY_PRICE]),
					timestamp: Number(res[0][CST.DB_HISTORY_TIMESTAMP]),
					volume: Number(res[0][CST.DB_HISTORY_VOLUME])
			}
			: { price: 0, timestamp: 0, volume: 0 };
	}

	async readSourceData(currentTimestamp: number): Promise<Trade[]> {
		const lowerTime = currentTimestamp - 3600000 + '';
		const upperTime = currentTimestamp + '';
		const res: object[] = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_TABLE_TRADE +
				' WHERE ' +
				CST.DB_TX_TS +
				' >= ' +
				lowerTime +
				' AND ' +
				CST.DB_TX_TS +
				' <= ' +
				upperTime
		);
		return res.map(item => ({
			source: item[CST.DB_TX_SRC],
			id: item[CST.DB_TX_ID],
			price: Number(item[CST.DB_TX_PRICE]),
			amount: Number(item[CST.DB_TX_AMOUNT]),
			timestamp: Number(item[CST.DB_TX_TS])
		}));
	}
}

const sqlUtil = new SqlUtil();
export default sqlUtil;
