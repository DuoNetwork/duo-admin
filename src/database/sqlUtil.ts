import * as mysql from 'mysql';
import * as CST from '../constants';
import { IPrice, ITrade } from '../types';
import util from '../util';
import dynamoUtil from './dynamoUtil';

export class SqlUtil {
	private conn: undefined | mysql.Connection = undefined;
	public init(host: string, user: string, pwd: string) {
		this.conn = mysql.createConnection({
			host: host,
			user: user,
			password: pwd,
			database: CST.DB_SQL_SCHEMA_PRICEFEED
		});

		this.conn.connect(err => {
			if (err) throw err;
			util.log('Connected!');
		});
	}

	public executeQuery(sqlQuery: string): Promise<any> {
		// util.log(sqlQuery);
		return new Promise((resolve, reject) => {
			if (this.conn)
				this.conn.query(sqlQuery, (err, result) => {
					if (err && err.code !== undefined && err.code === 'ER_DUP_ENTRY')
						// util.log('.');
						// rocess.stdout.write(".");
						reject(err);
					else if (err) reject(err);
					else resolve(result);
				});
			else reject('sql db connection is not initialized');
		});
	}

	public async insertTradeData(trade: ITrade, insertStatus: boolean) {
		const systemTimestamp = util.getNowTimestamp(); // record down the MTS

		// To do the checking for out of boundary data.
		// if(isNaN(trade.price)){
		// 	util.log('Price is NaN!');
		// 	return;
		// }

		// if(isNaN(trade.amount)){
		// 	util.log('Amount is NaN!');
		// 	return;
		// }

		// let price_str = math.format(price, { exponential: { lower: 1e-100, upper: 1e100 } });
		// let amount_str = math.format(amount, { exponential: { lower: 1e-100, upper: 1e100 } });

		const priceStr = trade.price.toString();
		const amountStr = trade.amount.toString();

		const sql =
			'REPLACE ' +
			CST.DB_SQL_TRADE +
			" VALUES ('" +
			trade.source +
			"','" +
			trade.id +
			"','" +
			priceStr +
			"','" +
			amountStr +
			"','" +
			(trade.timestamp || systemTimestamp) +
			"','" +
			systemTimestamp +
			"')";
		// util.log(await this.executeQuery(sql));
		await this.executeQuery(sql);
		if (insertStatus)
			await dynamoUtil.insertStatusData(
				dynamoUtil.convertTradeToSchema(trade, systemTimestamp)
			);
	}

	public async insertPrice(price: IPrice) {
		util.log(
			await this.executeQuery(
				'INSERT INTO ' +
					CST.DB_SQL_HISTORY +
					" VALUES ('" +
					price.timestamp +
					"','" +
					price.price +
					"','" +
					price.volume +
					"')"
			)
		);
		await dynamoUtil.insertStatusData(dynamoUtil.convertPriceToSchema(price));
	}

	public async readLastPrice(): Promise<IPrice> {
		const res = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_HISTORY +
				' order by ' +
				CST.DB_HISTORY_TIMESTAMP +
				' DESC LIMIT 1'
		);
		return res[0]
			? {
					price: Number(res[0][CST.DB_HISTORY_PRICE]),
					timestamp: Number(res[0][CST.DB_HISTORY_TIMESTAMP]),
					volume: Number(res[0][CST.DB_HISTORY_VOLUME])
			}
			: { price: 0, timestamp: 0, volume: 0 };
	}

	public async readSourceData(currentTimestamp: number): Promise<ITrade[]> {
		const lowerTime = currentTimestamp - 3600000 + '';
		const upperTime = currentTimestamp + '';
		const res: object[] = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_TRADE +
				' WHERE ' +
				CST.DB_TX_TS +
				' >= ' +
				lowerTime +
				' AND ' +
				CST.DB_TX_TS +
				' <= ' +
				upperTime
		);
		return res.map(item => ({
			source: item[CST.DB_TX_SRC],
			id: item[CST.DB_TX_ID],
			price: Number(item[CST.DB_TX_PRICE]),
			amount: Number(item[CST.DB_TX_AMOUNT]),
			timestamp: Number(item[CST.DB_TX_TS])
		}));
	}
}

const sqlUtil = new SqlUtil();
export default sqlUtil;
