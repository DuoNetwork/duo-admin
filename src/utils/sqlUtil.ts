import * as mysql from 'mysql';
import * as CST from '../common/constants';
import { IPriceFix, ITrade } from '../common/types';
import dynamoUtil from './dynamoUtil';
import util from './util';

class SqlUtil {
	private conn: undefined | mysql.Connection = undefined;
	public init(host: string, user: string, pwd: string) {
		this.conn = mysql.createConnection({
			host: host,
			user: user,
			password: pwd,
			database: CST.DB_SQL_SCHEMA_PRICEFEED
		});
		this.conn.on('error', err => {
			if (err.code === 'PROTOCOL_CONNECTION_LOST') {
				util.logInfo('ERROR: Server Disconnects. Reconnecting');
				this.init(host, user, pwd);
			} else throw err;
		});

		return new Promise<void>((resolve, reject) => {
			if (this.conn)
				this.conn.connect(err => {
					if (err) reject(err);
					util.logInfo('Connected!');
					resolve();
				});
			else reject('no connection');
		});
	}

	public executeQuery(sqlQuery: string): Promise<any> {
		// util.logInfo(sqlQuery);
		return new Promise((resolve, reject) => {
			if (this.conn)
				this.conn.query(sqlQuery, (err, result) => {
					if (err && err.code !== undefined && err.code === 'ER_DUP_ENTRY')
						// util.logInfo('.');
						// rocess.stdout.write(".");
						reject(err);
					else if (err) reject(err);
					else resolve(result);
				});
			else reject('sql db connection is not initialized');
		});
	}

	public getTradeCounts() {
		const sql = `SELECT COUNT(*) FROM ${CST.DB_SQL_TRADE} WHERE 1`;
		return this.executeQuery(sql);
	}

	public async insertTradeData(trade: ITrade, insertStatus: boolean) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS

		// To do the checking for out of boundary data.
		// if(isNaN(trade.price)){
		// 	util.logInfo('Price is NaN!');
		// 	return;
		// }

		// if(isNaN(trade.amount)){
		// 	util.logInfo('Amount is NaN!');
		// 	return;
		// }

		// let price_str = math.format(price, { exponential: { lower: 1e-100, upper: 1e100 } });
		// let amount_str = math.format(amount, { exponential: { lower: 1e-100, upper: 1e100 } });

		const priceStr = trade.price.toString();
		const amountStr = trade.amount.toString();

		const sql =
			'REPLACE ' +
			CST.DB_SQL_TRADE +
			" VALUES ('" +
			trade.source +
			"','" +
			trade.id +
			"','" +
			priceStr +
			"','" +
			amountStr +
			"','" +
			(trade.timestamp || systemTimestamp) +
			"','" +
			systemTimestamp +
			"','" +
			(trade.quote + '|' + trade.base) +
			"')";
		// util.logInfo(await this.executeQuery(sql));
		await this.executeQuery(sql);
		if (insertStatus)
			await dynamoUtil.insertStatusData(
				dynamoUtil.convertTradeToDynamo(trade, systemTimestamp)
			);
	}

	public async insertPrice(price: IPriceFix) {
		util.logInfo(
			await this.executeQuery(
				'INSERT INTO ' +
					CST.DB_SQL_HISTORY +
					" VALUES ('" +
					price.timestamp +
					"','" +
					price.price +
					"','" +
					price.volume +
					"','" +
					(price.quote + '|' + price.base) +
					"')"
			)
		);
	}

	public async readLastPrice(base: string, quote: string): Promise<IPriceFix> {
		const pair = quote + '|' + base;
		const res = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_HISTORY +
				' order by ' +
				CST.DB_HISTORY_TIMESTAMP +
				` WHERE pair = '${pair}' DESC LIMIT 1`
		);
		return res[0]
			? {
					price: Number(res[0][CST.DB_HISTORY_PRICE]),
					timestamp: Number(res[0][CST.DB_HISTORY_TIMESTAMP]),
					volume: Number(res[0][CST.DB_HISTORY_VOLUME]),
					source: '',
					base: base,
					quote: quote
			}
			: {
					price: 0,
					timestamp: 0,
					volume: 0,
					source: '',
					base: base,
					quote: quote
			};
	}

	public async readSourceData(
		currentTimestamp: number,
		base: string,
		quote: string
	): Promise<ITrade[]> {
		const pair = quote + '|' + base;
		const lowerTime = currentTimestamp - 3600000 + '';
		const upperTime = currentTimestamp + '';
		const res: Array<{ [key: string]: string }> = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_TRADE +
				' WHERE ' +
				CST.DB_TX_TS +
				' >= ' +
				lowerTime +
				' AND ' +
				CST.DB_TX_TS +
				' <= ' +
				upperTime +
				` AND pair = '${pair}';`
		);
		return res.map(item => ({
			quote: item[CST.DB_TX_QTE],
			base: item[CST.DB_TX_BASE],
			id: item[CST.DB_TX_ID],
			source: item[CST.DB_TX_SRC],
			price: Number(item[CST.DB_TX_PRICE]),
			amount: Number(item[CST.DB_TX_AMOUNT]),
			timestamp: Number(item[CST.DB_TX_TS])
		}));
	}

	public async cleanDB(): Promise<void> {
		const queryString =
			'DELETE FROM ' +
			CST.DB_SQL_TRADE +
			' WHERE timestamp < UNIX_TIMESTAMP(DATE_SUB(NOW(), INTERVAL 7 DAY)) * 1000';
		util.logInfo(queryString);
		this.executeQuery(queryString);
		const res = await this.getTradeCounts();
		util.logDebug(`num of trades existing in db ${JSON.stringify(res)}`);
	}
}

const sqlUtil = new SqlUtil();
export default sqlUtil;
