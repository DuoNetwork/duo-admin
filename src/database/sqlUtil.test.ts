import { IPrice } from '../types';
import dynamoUtil from './dynamoUtil';
import sqlUtil from './sqlUtil';

test('connection initalization', () => {
	return sqlUtil.executeQuery('test').catch(error => expect(error).toMatchSnapshot());
});

test('insertTradeData', async () => {
	sqlUtil.executeQuery = jest.fn(() => Promise.resolve());
	dynamoUtil.insertData = jest.fn(() => Promise.resolve());
	Date.now = jest.fn(() => 0);
	await sqlUtil.insertTradeData({
		source: 'src',
		id: 'id',
		price: 123,
		amount: 456,
		timestamp: 1234567890
	});
	expect((sqlUtil.executeQuery as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});

test('insertPrice', async () => {
	sqlUtil.executeQuery = jest.fn(() => Promise.resolve());
	dynamoUtil.insertData = jest.fn(() => Promise.resolve());
	await sqlUtil.insertPrice({
		price: 123,
		volume: 456,
		timestamp: 1234567890
	});
	expect((sqlUtil.executeQuery as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});

test('readLastPrice', async () => {
	sqlUtil.executeQuery = jest.fn(() =>
		Promise.resolve([
			{
				price: '123',
				volume: '456',
				timestamp: '1234567890'
			}
		])
	);
	const price = await sqlUtil.readLastPrice();
	expect(
		(sqlUtil.executeQuery as jest.Mock<Promise<IPrice[]>>).mock.calls[0][0]
	).toMatchSnapshot();
	expect(price).toMatchSnapshot();
});

test('readSourceData', async () => {
	sqlUtil.executeQuery = jest.fn(() =>
		Promise.resolve([
			{
				source: 'src1',
				id: 'id1',
				price: '123',
				amount: '456',
				timestamp: '1234567890'
			},
			{
				source: 'src2',
				id: 'id2',
				price: '234',
				amount: '567',
				timestamp: '2345678901'
			}
		])
	);
	const trades = await sqlUtil.readSourceData(1234567890);
	expect((sqlUtil.executeQuery as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect(trades).toMatchSnapshot();
});
