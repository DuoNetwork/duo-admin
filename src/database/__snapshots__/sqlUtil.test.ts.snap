// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connection initalization 1`] = `"sql db connection is not initialized"`;

exports[`insertPrice 1`] = `"INSERT INTO eth_historical_price VALUES ('1234567890','123','456')"`;

exports[`insertTradeData 1`] = `"REPLACE eth_trades VALUES ('src','id','123','456','1234567890','0')"`;

exports[`readLastPrice 1`] = `"SELECT * FROM eth_historical_price order by timestamp DESC LIMIT 1"`;

exports[`readLastPrice 2`] = `
Object {
  "price": 123,
  "timestamp": 1234567890,
  "volume": 456,
}
`;

exports[`readSourceData 1`] = `"SELECT * FROM eth_trades WHERE timestamp >= 1230967890 AND timestamp <= 1234567890"`;

exports[`readSourceData 2`] = `
Array [
  Object {
    "amount": 456,
    "id": "id1",
    "price": 123,
    "source": "src1",
    "timestamp": 1234567890,
  },
  Object {
    "amount": 567,
    "id": "id2",
    "price": 234,
    "source": "src2",
    "timestamp": 2345678901,
  },
]
`;
